[DO-518] How to send payment with Gevamu (#4)
gevamu/gevamu.github.io: content/4.examples/send-payment.md

    return flowHandle.returnValue.get()
  }
}
```

## Transactions lifecycle

Once payment instruction is received, the Gevamu Payment Solution validates and authorizes (by checking node identity and the Participant ID) the payment.
If the validation passes, the Payment Gateway sets the payment status to ‘Sent to Gateway’. Status change is communicated back to the Participant's node.
Transaction can be accepted or rejected by the Payment Service Provider.

To make a transaction as a network participant, describe Payment in one of formats listed in [`PaymentInstructionFormat`](https://gevamu.github.io/corda-payments-sdk/payments-workflows/com.gevamu.corda.flows/-payment-instruction-format/index.html) enum.

You can find sequence diagrams of payment processing and payment state transition below:

```mermaid

sequenceDiagram
  participant CA as MyPaymentService
  participant PCD as GevamuFacade

  CA->>PCD: Initiate Payment
  PCD->>PCD: Initiate payment <br/> within PSP
  PCD->>CA: Payment status: Created
  PCD->>PCD: Passive payment update...
  Note over CA,PCD: After state change
  CA->>PCD: Query payment
  PCD->>CA: Payment with relevant status

```

```mermaid

> stateDiagram-v2

Reviewer: Add full name for Sent status.
```suggestion
stateDiagram-v2
    Sent: Sent to gateway
```

Author: Is it an output are you going to have, @d0rich ? 
![image](https://github.com/gevamu/gevamu.github.io/assets/62986513/54fa6e55-65dd-4cf3-a3e0-037becb2e3ea)

Reviewer: Nope, this output:

![image](https://github.com/gevamu/gevamu.github.io/assets/51794590/6e4e0642-94bf-4f70-984c-a3c269663433)

Full code:

```mermaid
stateDiagram-v2
    Sent: Sent to gateway
    [*] --> Created
    Created --> Sent
    Sent --> Pending
    Pending --> Accepted
    Pending --> Rejected
    Accepted --> [*]
    Rejected --> [*]
```

# Send Payment

This example shows how to send a payment from one account to another via Gevamu Payments Solution.

```mermaid
classDiagram

>   class MyPaymentService {

Reviewer 2: Do we need `MyPaymentService` in this example?

    + sendPaymentViaGevamu()
  }
  class GevamuFacade {
    + sendPayment()
  }

  MyPaymentService ..> GevamuFacade
```

## Program implementation

Sending payment is quite simple operation. It includes only call of `PaymentFlow`. 

But before starting the flow you should provide payment instruction. It should be created according to chosen payment standard.

```kotlin
class MyPaymentService(private val gevamuFacade: GevamuFacade) {
    fun sendPaymentViaGevamu() {
        // Create payment instruction with payment details
        val paymentInstruction = PaymentInstruction(/* payment details */)
        
        // Specify gateway node for payment processing
        val gateway = Party(/* gateway node data */)
        
        // Use GevamuFacade to send payment instruction to gateway node
        gevamuFacade.sendPayment(paymentInstruction, gateway)
    }
}

class GevamuFacade(private val serviceHub: AppServiceHub): SingletonSerializeAsToken() {

  fun sendPayment(
    paymentInstruction: PaymentInstruction,
    gateway: Party

>   ): List<StateAndRef<Payment>> {

Reviewer 2: It's worth explaining why `List` is returned from the `PaymentFlow`

Author: Explanation added

>     val flowHandle = serviceHub.startFlow(PaymentFlow(paymentInstruction, gateway, UUID.randomUUID()))

Reviewer 2: If I'm not mistaken, the last parameter (UUID) is optional. If it is I'd remove it in this example.

Author: That's true, will remove it in the next commit

    return flowHandle.returnValue.get()
  }
}
```

## Transactions lifecycle

Once payment instruction is received, the Gevamu Payment Solution validates and authorizes (by checking node identity and the Participant ID) the payment.

> If the validation passes, the Payment Gateway sets the payment status to ‘Sent to Gateway’. Status change is communicated back to the Participant's node.

Reviewer 2: It relates to my previous comment. If payment initiation request has multiple payment instructions, multiple payment states are created and are independently updated.

Author: Explanation added

> Transaction can be accepted or rejected by the Payment Service Provider.

Reviewer 2: If payment cannot be immediately completed, "Pending" status is returned. The payment state will be updated with the final status when the Gevamu gateway receives new status from the PSP.

Author: Explanation added

> To make a transaction as a network participant, describe Payment in one of formats listed in [`PaymentInstructionFormat`](https://gevamu.github.io/corda-payments-sdk/payments-workflows/com.gevamu.corda.flows/-payment-instruction-format/index.html) enum.

Reviewer 2: For me this phrase implies that it's possible to make a transaction not as a network participant :-)

Author: Fixed in next commit

You can find sequence diagrams of payment processing and payment state transition below:

```mermaid

sequenceDiagram
  participant CA as MyPaymentService
  participant PCD as GevamuFacade

  CA->>PCD: Initiate Payment
  PCD->>PCD: Initiate payment <br/> within PSP
  PCD->>CA: Payment status: Created
  PCD->>PCD: Passive payment update...
  Note over CA,PCD: After state change
  CA->>PCD: Query payment
  PCD->>CA: Payment with relevant status

```

```mermaid

stateDiagram-v2
    Sent: Sent to gateway

>     [*] --> Created
>     Created --> Sent
>     Sent --> Pending
>     Pending --> Accepted
>     Pending --> Rejected
>     Accepted --> [*]
>     Rejected --> [*]

Reviewer 2: ```suggestion
    [*] --> Created
    Created --> Sent
    Sent --> Pending
    Sent  --> Accepted
    Sent  --> Rejected
    Pending --> Accepted
    Pending --> Rejected
    Accepted --> [*]
    Rejected --> [*]
```

# Send Payment

This example shows how to send a payment from one account to another via Gevamu Payments Solution.

```mermaid
classDiagram
  class GevamuFacade {
    + sendPayment()
  }
```

## Program implementation

> Sending payment is quite simple operation. It includes only call of `PaymentFlow`. 

Reviewer 3: subjective statement - please rephrase

Reviewer 3: ```suggestion
Sending a payment is a one-step operation, consisting of a single `PaymentFlow` call. 
```

## Transactions lifecycle

Once payment instruction is received, the Gevamu Payment Solution validates and authorizes (by checking node identity and the Participant ID) the payment.
If the validation passes, the Payment Gateway sets the payment status to ‘Sent to Gateway’. 
Status change is communicated back to the Participant's node.
Some payment initiation requests has multiple payment instructions.

> In this case multiple payment states are created and independently updated.

Reviewer 3: If a payment initiation request contains multiple payment instructions, multiple payment states will be created and independently updated. 
(singular/plural nouns should be grammatically aligned with the verb)

> Some payment initiation requests has multiple payment instructions.
> In this case multiple payment states are created and independently updated.

Author: ```suggestion
If payment initiation requests contain multiple payment instructions, multiple payment states will be created and independently updated.
```
Now better?

> Some payment initiation requests has multiple payment instructions.
> In this case multiple payment states are created and independently updated.

Reviewer 3: ```suggestion
If a payment initiation request contains multiple payment instructions, multiple payment states will be created and independently updated.
```
